### PR TITLE
Avoid Operator construction overhead in Optimize 1q decomposition pass

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit further tunes the performance of the optimize 1q
decomposition pass. Previously, the pass was constructing an operator
using the Operator qi class and then calling compose() on each gate in
the run to find the overall unitary of the run. However, doing this was
constructing a lot of intermediate operators that weren't needed and
just added extra overhead. To avoid that overhead this commit relies on
the raw numpy calculations to build the matrix for the overall operator
and then wraps that in an Operator() object only once when we go to call
the OneQubitEulerDecomposer.

### Details and comments

### CODE REVIEW

**Review by kdk — COMMENTED** [review]
I think this looks good, and will likely go a long way to making `Optimize1qGateDecompositions` more comparable in performance to `Optimize1qGates`. Do you have any preliminary performance numbers?

File: qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py [hunk 1]
```diff
@@ -91,11 +91,11 @@ def run(self, dag):
                     continue
 
             new_circs = []
-            operator = Operator(run[0].op)
+            operator = np.array(run[0].op, dtype=complex)
             for gate in run[1:]:
-                operator = operator.compose(gate.op)
+                operator = np.dot(gate.op, operator)
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py:96]: Really, that's surprising, `to_matrix()` should just be calling `__array__`: https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/circuit/gate.py#L55-L57

[the remark this answers is not in the document]
Comment (kdk) [reply qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py:96]: It does, but I think the other way around is more costly (numpy having to inspect to see if an `__array__` exists, and then subsequently calling it, at a guess).

At least locally, I see:

```
%%timeit r = [HGate() if _ % 2 == 0 else XGate() for _ in range(5)]
s = Operator(r[0])
for t in r[1:]:
    s = s.compose(t)

243 µs ± 39.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

%%timeit r = [HGate() if _ % 2 == 0 else XGate() for _ in range(5)]
s = np.array(r[0], dtype=complex)
for t in r[1:]:
    s = np.dot(t, s)

86 µs ± 3.53 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

%%timeit r = [HGate() if _ % 2 == 0 else XGate() for _ in range(5)]
s = r[0].to_matrix()
for t in r[1:]:
    s = np.dot(t.to_matrix(), s)

42.2 µs ± 2.19 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

%%timeit r = [HGate() if _ % 2 == 0 else XGate() for _ in range(5)]
s = r[0].to_matrix()
for t in r[1:]:
    s = t.to_matrix().dot(s)

40.6 µs ± 3.85 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

**Comment by mtreinish** [comment]
I did some quick manual benchmarking (only one sample each run so not sure what the error is) with a `random_circuit(65, 4096, measure=True, seed=42)` run through the basis translator first the pass takes:

| branch | basis | time (sec) |
| --------- | ------- | ------ |
| master | aer's | 35.0315 |
| master | new ibmq | 29.5729 |
| This PR | aer's | 33.1140 |
| This PR | new ibmq | 25.8476 |
| w/ to_matrix() | aer's | 33.4808 |
| w/ to_matrix() | new ibmq | 24.9945 |
| w/ to_matrix() and operator.dot() | aer's | 33.4753 |
| w/ to_matrix() and operator.dot() | new ibmq | 24.8618 |

From when I profiled the pass recently: https://github.com/Qiskit/qiskit-terra/issues/5775#issuecomment-777807475 the biggest bottleneck is the decomposer, so it looks like the 2x difference, which is only ~40us, ends up lost in the noise.

File: qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py [hunk 2]
```diff
@@ -91,11 +91,11 @@ def run(self, dag):
...
             for decomposer in self.basis:
-                new_circs.append(decomposer(operator))
+                new_circs.append(decomposer(Operator(operator)))
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py]: Ah yep good call, the decomposer is just going to pull the inner matrix out anyway. Done in: https://github.com/Qiskit/qiskit-terra/pull/5909/commits/e3618a84498bfc09cd03981666bcf8b94ab467aa

File: qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py]: ok, I updated this in https://github.com/Qiskit/qiskit-terra/pull/5909/commits/87403d883932deadee2eac99d62c071c8e83af49 40us doesn't really show up compared to the other parts of the pass, but going with what's fastest makes sense.

**Review by kdk — APPROVED** [review]

**Pull request merged**